src/solvers/primal_dual_semismooth_Newton.jl

```diff
@@ -0,0 +1,378 @@
+    update_primal_base::Union{Function,Missing}=missing,
+    update_dual_base::Union{Function,Missing}=missing,
+    retraction_method::RM=ExponentialRetraction(),
+    inverse_retraction_method::IRM=LogarithmicInverseRetraction(),
+    vector_transport_method::VTM=ParallelTransport(),
+    return_options=false,
+    kwargs...,
+) where {
+    mT<:AbstractManifold,
+    nT<:AbstractManifold,
+    P,
+    Q,
+    T,
+    RM<:AbstractRetractionMethod,
+    IRM<:AbstractInverseRetractionMethod,
+    VTM<:AbstractVectorTransportMethod,
+}
+    p = PrimalDualSemismoothNewtonProblem(
+        M,
+        N,
+        cost,
+        prox_F,
+        diff_prox_F,
+        prox_G_dual,
+        diff_prox_G_dual,
+        linearized_forward_operator,
+        adjoint_linearized_operator;
+        Λ=Λ,
+        evaluation=evaluation,
+    )
+    o = PrimalDualSemismoothNewtonOptions(
+        m,
+        n,
+        x,
+        ξ,
+        primal_stepsize,
+        dual_stepsize,
+        reg_param;
+        stopping_criterion=stopping_criterion,
+        update_primal_base=update_primal_base,
+        update_dual_base=update_dual_base,
+        retraction_method=retraction_method,
+        inverse_retraction_method=inverse_retraction_method,
+        vector_transport_method=vector_transport_method,
+    )
+    o = decorate_options(o; kwargs...)
+    resultO = solve(p, o)
+    if return_options
+        return resultO
+    else
+        return get_solver_result(resultO)
+    end
+end
+
+function initialize_solver!(
+    ::PrimalDualSemismoothNewtonProblem, ::PrimalDualSemismoothNewtonOptions
+) end
+
+function step_solver!(
+    p::PrimalDualSemismoothNewtonProblem, o::PrimalDualSemismoothNewtonOptions, iter
+)
+    # do step
+    primal_dual_step!(p, o)
+    o.m = ismissing(o.update_primal_base) ? o.m : o.update_primal_base(p, o, iter)
+    if !ismissing(o.update_dual_base)
+        n_old = deepcopy(o.n)
+        o.n = o.update_dual_base(p, o, iter)
+        o.ξ = vector_transport_to(p.N, n_old, o.ξ, o.n, o.vector_transport_method)
+        o.ξbar = vector_transport_to(p.N, n_old, o.ξbar, o.n, o.vector_transport_method)
+    end
+    return o
+end
+
+function primal_dual_step!(
+    p::PrimalDualSemismoothNewtonProblem, o::PrimalDualSemismoothNewtonOptions
+)
+
+    # construct X
+    X = construct_vector(p, o)
+
+    # construct matrix
+    ∂X = construct_matrix(p, o)
+    ∂X += o.reg_param * sparse(I, size(∂X))  # prevent singular matrix at solution
+
+    # solve matrix -> find coordinates
+    d_coords = ∂X \ -X
+
+    dims = manifold_dimension(p.M)
+    dx_coords = d_coords[1:dims]
+    dξ_coords = d_coords[(dims + 1):end]
+
+    # compute step
+    dx = get_vector(p.M, o.x, dx_coords, DefaultOrthonormalBasis())
+    dξ = get_vector(p.N, o.n, dξ_coords, DefaultOrthonormalBasis())
+
+    # do step
+    o.x = retract(p.M, o.x, dx, o.retraction_method)
+    return o.ξ = o.ξ + dξ
+end
+
+function construct_vector(
```

Comment:
could you choose here (and for the next function) a little nicer function names? This function name tells us that the function constructs _soma_ vector but it is completely unclear which one and what for. Though this is an internal helper function, it should still get a doc-string explaining the function a little mode.